lib/algos/list/delimiterOrphan.js

```diff
@@ -0,0 +1,175 @@
+
+                lastModified = v['last-modified'];
+                isDeleteMarker = v.isDeleteMarker;
+            } catch (e) {
+                this.logger.warn('could not parse Object Metadata while listing',
+                    {
+                        method: 'getLastModified',
+                        err: e.toString(),
+                    });
+            }
+            // We then check if the orphan version is a delete marker and if it is older than the "beforeDate"
+            if ((!this.beforeDate || (lastModified && lastModified < this.beforeDate)) && isDeleteMarker) {
+                this.Contents.push({ key: this.keyName, value: this.trimMetadata(this.value) });
+                this.NextMarker = this.keyName;
+                ++this.keys;
+            }
+        }
+    }
+
+    /**
+     * NOTE: Each version of a specific key is sorted from the latest to the oldest
+     * thanks to the way version ids are generated.
+     * DESCRIPTION: For a given key, the latest version is kept in memory since it is the current version.
+     * If the following version reference a new key, it means that the previous one was an orphan version.
+     * We then check if the orphan version is a delete marker and if it is older than the "beforeDate"
+     * The process stops and returns the available results if either:
+     * - no more metadata key is left to be processed
+     * - the listing reaches the maximum number of key to be returned
+     * - the internal timeout is reached
+     * NOTE: we cannot leverage MongoDB to list keys older than "beforeDate"
+     * because then we will not be able to assess its orphanage.
+     *  @param {String} keyVersionSuffix   - The key with version id as a suffix.
+     *  @param {String} value              - The value of the key
+     *  @return {number}                   - indicates if iteration should continue
+     */
+    addContents(keyVersionSuffix, value) {
+        if (this._reachedMaxKeys()) {
+            return FILTER_END;
+        }
+
+        if (this.start && Date.now() - this.start > DELIMITER_TIMEOUT_MS) {
+            this.IsTruncated = true;
+            this.NextMarker = this.keyName;
+
+            this.logger.info('listing stopped after expected internal timeout',
+                {
+                    timeoutMs: DELIMITER_TIMEOUT_MS,
+                    evaluatedKeys: this.evaluatedKeys,
+                });
+            return FILTER_END;
+        }
+        ++this.evaluatedKeys;
+
+        const versionIdIndex = keyVersionSuffix.indexOf(VID_SEP);
+        // key without version suffix
+        const key = keyVersionSuffix.slice(0, versionIdIndex);
+
+        // For a given key, the youngest version is kept in memory since it represents the current version.
+        if (key !== this.keyName) {
+            // If the following version reference a new key,
+            // it means that the previous one (this.value) was an orphan version.
+            this._addOrphan();
```

Comment:
I had a hard time understanding why this would always "swallow" the first old enough delete marker that passes through, as the `if (this.value)` test in `_addOrphan` looks more like a regular null-check than an orphan-state test.

For clarity I'd remove the `if (this.value)` test from `_addOrphan` (to clarify its purpose), and instead add it here, with a comment saying that if `this.value != null` it means that the key's not the first to be evaluated and this is "allowed" to be an orphan.